pkg/expression/builtin_cast.go

```diff
@@ -2550,6 +2550,9 @@ func WrapWithCastAsInt(ctx BuildContext, expr Expression) Expression {
 	tp.SetDecimal(0)
 	types.SetBinChsClnFlag(tp)
 	tp.AddFlag(expr.GetType(ctx.GetEvalCtx()).GetFlag() & (mysql.UnsignedFlag | mysql.NotNullFlag))
+	if targetType != nil {
+		tp.AddFlag(targetType.GetFlag() & mysql.UnsignedFlag)
+	}
```

Comment:
```suggestion
    // inherit NotNullFlag from source type
    tp.AddFlag(expr.GetType(ctx.GetEvalCtx()).GetFlag() & mysql.NotNullFlag)
    if targetType == nil {
        // inherit UnsignedFlag from source type if targetType is nil
        tp.AddFlag(expr.GetType(ctx.GetEvalCtx()).GetFlag() & mysql.UnsignedFlag)
    } else {
        // otherwise set UnsignedFlag based on targetType
		tp.AddFlag(targetType.GetFlag() & mysql.UnsignedFlag)
	}
```